TWTValidation/JSON Validator/AST Nodes/TWTJSONSchemaASTNode.m

```diff
@@ -43,5 +43,111 @@ - (NSSet *)validTypes
+    [referenceNodes addObjectsFromArray:self.notSchema.childrenReferenceNodes];
+
+    [referenceNodes addObjectsFromArray:[self childrenReferenceNodesFromNodeDictionary:self.definitions]];
+
+    return referenceNodes;
+}
+
+
+- (NSArray *)childrenReferenceNodesFromNodeArray:(NSArray *)array
+{
+    if (!array) {
+        return @[ ];
+    }
+
+    NSMutableArray *referenceNodes = [[NSMutableArray alloc] init];
+    for (TWTJSONSchemaASTNode *node in array) {
+        [referenceNodes addObjectsFromArray:node.childrenReferenceNodes];
+    }
+    return [referenceNodes copy];
+}
+
+
+- (NSArray *)childrenReferenceNodesFromNodeDictionary:(NSDictionary *)dictionary
+{
+    if (!dictionary) {
+        return @[ ];
+    }
+
+    NSMutableArray *referenceNodes = [[NSMutableArray alloc] init];
+    for (NSString *key in dictionary) {
+        [referenceNodes addObjectsFromArray:[dictionary[key] childrenReferenceNodes]];
+    }
+    return [referenceNodes copy];
+}
+
+
+- (TWTJSONSchemaASTNode *)nodeForPathComponents:(NSArray *)path
+{
+    if (path.count == 0) {
+        return self;
+    }
+
+    NSString *key = path.firstObject;
+    NSArray *remainingPath = [self remainingPathFromPath:path];
+
+    if ([key isEqualToString:TWTJSONSchemaKeywordAllOf]) {
+        return [self nodeForPathComponents:remainingPath fromNodeArray:self.andSchemas];
+    }
+    if ([key isEqualToString:TWTJSONSchemaKeywordAnyOf]) {
+        return [self nodeForPathComponents:remainingPath fromNodeArray:self.orSchemas];
+    }
+    if ([key isEqualToString:TWTJSONSchemaKeywordOneOf]) {
+        return [self nodeForPathComponents:remainingPath fromNodeArray:self.exactlyOneOfSchemas];
+    }
+    if ([key isEqualToString:TWTJSONSchemaKeywordNot]) {
+        return [self.notSchema nodeForPathComponents:remainingPath];
+    }
+    if ([key isEqualToString:TWTJSONSchemaKeywordDefinitions]) {
+        return [self nodeForPathComponents:remainingPath fromDefinitions:self.definitions];
+    }
+
+    return nil;
+}
+
+
+- (TWTJSONSchemaASTNode *)nodeForPathComponents:(NSArray *)path fromNodeArray:(NSArray *)array
+{
+    NSString *index = path.firstObject;
+    if (!index || ![self componentIsIndex:index] || index.integerValue > array.count) {
+        return nil;
+    }
+
+    return [array[index.integerValue] nodeForPathComponents:[self remainingPathFromPath:path]];
+}
+
+
+- (TWTJSONSchemaASTNode *)nodeForPathComponents:(NSArray *)path fromDefinitions:(NSDictionary *)definitions
+{
+    NSString *key = path.firstObject;
+
+    return [definitions[key] nodeForPathComponents:[self remainingPathFromPath:path]];
+}
+
+
+- (BOOL)componentIsIndex:(NSString *)key
+{
+    return [key rangeOfCharacterFromSet:[[NSCharacterSet decimalDigitCharacterSet] invertedSet]].location == NSNotFound;
+}
+
+
+- (NSArray *)remainingPathFromPath:(NSArray *)path
+{
+    NSRange range;
+    range.location = 1;
+    range.length = path.count - 1;
+    return [path subarrayWithRange:range];
```

Comment:
Using `NSMakeRange(NSUInteger location, NSUInteger length)` that will make this a one-line method.